src/Providers/Http/Util/ResponseUtil.php

```diff
@@ -15,53 +18,54 @@
+     * Throws an appropriate exception if the given response is not successful.
      *
      * This method checks the HTTP status code of the response and throws
-     * a ResponseException if the status code indicates an error (i.e., not in the
-     * 2xx range). It also attempts to extract a more detailed error message from
-     * the response body if available.
+     * the appropriate exception type based on the status code range:
+     * - 3xx: RedirectException (redirect responses)
+     * - 4xx: ClientException (client errors)
+     * - 5xx: ServerException (server errors)
+     * - Other unsuccessful responses: ResponseException (malformed responses)
      *
      * @since 0.1.0
      *
      * @param Response $response The HTTP response to check.
-     * @throws ResponseException If the response is not successful.
+     * @throws RedirectException If the response indicates a redirect (3xx).
+     * @throws ClientException If the response indicates a client error (4xx).
+     * @throws ServerException If the response indicates a server error (5xx).
+     * @throws ResponseException If the response format is unexpected.
      */
     public static function throwIfNotSuccessful(Response $response): void
     {
         if ($response->isSuccessful()) {
             return;
         }
 
-        $errorMessage = sprintf(
-            'Bad status code: %d.',
-            $response->getStatusCode()
-        );
+        $statusCode = $response->getStatusCode();
 
-        // Handle common error formats in API responses.
-        $data = $response->getData();
-        if (
-            is_array($data) &&
-            isset($data['error']) &&
-            is_array($data['error']) &&
-            isset($data['error']['message']) &&
-            is_string($data['error']['message'])
-        ) {
-            $errorMessage .= ' ' . $data['error']['message'];
-        } elseif (
-            is_array($data) &&
-            isset($data['error']) &&
-            is_string($data['error'])
-        ) {
-            $errorMessage .= ' ' . $data['error'];
-        } elseif (
-            is_array($data) &&
-            isset($data['message']) &&
-            is_string($data['message'])
-        ) {
-            $errorMessage .= ' ' . $data['message'];
+        // 3xx Redirect Responses
+        if ($statusCode >= 300 && $statusCode < 400) {
+            throw RedirectException::fromRedirectResponse($response);
         }
 
-        throw new ResponseException($errorMessage, $response->getStatusCode());
+        // 4xx Client Errors
+        if ($statusCode >= 400 && $statusCode < 500) {
+            // Special handling for 400 Bad Request
+            if ($statusCode === 400) {
+                $body = (string) $response->getBody();
+                $errorDetail = $body ? substr($body, 0, 200) : 'Invalid request parameters';
+                throw ClientException::fromBadRequestResponse($errorDetail);
+            }
```

Comment:
Why is this needed? When is the response body a simple string?

I think this should use the same logic as below just like the other 4xx error codes, unless I'm missing something?